### ISSUE TITLE
ppc64le and s390x arches build failed

### ISSUE THREAD

**Issue by Vascom** [sadko4u/lsp-plugins#52]
I want add lsp-plugins to Fedora but it failed build on ppc64le and s390x arches.
Errors in vst2.h
```
BUILDSTDERR: /builddir/build/BUILD/lsp-plugins-src-1.1.9/include/3rdparty/steinberg/vst2.h:2179:34: error: expected ')' before '*' token
BUILDSTDERR:  2179 | typedef VstIntPtr   (VSTCALLBACK *audioMasterCallback)      (AEffect* effect, VstInt32 opcode, VstInt32 index, VstIntPtr value, void* ptr, float opt);
BUILDSTDERR:       |                     ~            ^
```
Full build log with all errors https://kojipkgs.fedoraproject.org//work/tasks/4222/35834222/build.log

Can you help me?

**Comment by sadko4u** [comment]
Hello!

Currently the main problem is that LSP Plugins do support only several architectures - i586, x86_64.
Also as experimental architectures ARMv7 and AArch64 are supported.

To add support of ppc64le and s390x, we probably need to create a special branch in GIT repository and push commits step-by-step that will finally give us the plugins compiled. This will not be quick, though, but I'm actually interested in support of other architectures and portability improvements.

According to your current compilation error: the official Steinberg's headers that were taken as base of vst2.h header defined the VSTCALLBACK macro which, actually, is specified only for several platforms.

I've merged latest changest to **devel** branch, made **ppc64-port** sub-branch and committed first change: 0eede5922176ac7f7f2091881f1d5317514f1064

Please try to build now.

**Comment by Vascom** [comment]
Same errors https://kojipkgs.fedoraproject.org//work/tasks/4178/35854178/build.log
I am applied patch to 1.1.9 release. It is normal?

**Comment by sadko4u** [comment]
I think better to start with, is trying to build from the branch I've created directly. Then we'll backport all necessary changes into **hotfix** branch that you can use as a patch source to the 1.1.9 release.

Anyway, according to logs, you've got already a lot of code compiled, so there are not so many things left that should be changed in the code.

Also currently I see that **vst2.h** still hasn't detected the PPC64 architecture. I need to look at predefined compiler macros on your machine. Could you issue  'gcc -dM -E - < /dev/null' command on the builder's machine and post the result here?

**Comment by Vascom** [comment]
Predefined compiler macros:
https://paste.fedoraproject.org/paste/o44FmXX2GmQNWCc1kovxSA

**Comment by Vascom** [comment]
With ppc64-port branch the same errors
https://paste.fedoraproject.org/paste/o44FmXX2GmQNWCc1kovxSA

**Comment by sadko4u** [comment]
Thanks a lot for the list of predefined macros. Here's the patch that should test __PPC64__ and __LP64__ macros in order to detect the PowerPC architecture: 9bcb155f928b4b4fb058b85eb0f1d704b69e7d53

**Comment by Vascom** [comment]
Build successful.
You will create new release or give me patch for 1.1.9?

And now time for s390x?

**Comment by sadko4u** [comment]
Wow, that sounds nice. I would like also to ask you to launch unit tests to ensure that at least code covered by unit tests works fine. This may be done by issuing the following commands:
```
make test
.test/lsp-plugins-test utest
```
Also it would be nice if you provide the output of unit tests saved to the file:
```
.test/lsp-plugins-test utest >lsp-plugins-utest.log 2>&1
```

**Comment by Vascom** [comment]
Done
[lsp-plugins-utest.log](https://github.com/sadko4u/lsp-plugins/files/3338527/lsp-plugins-utest.log)

**Comment by sadko4u** [comment]
Thanks. Seems that we've found a problem when converting endianess of **double**-typed arguments. This patch should fix the problem: 102d37fdfb65d3eb987b4dbbd5d457e54b3af2d5
Please rebuild and relaunch. To not to mess output of different threads, please launch tests in single-thread mode:
```
.test/lsp-plugins-test utest -j 1 >lsp-plugins-utest.log 2>&1
```
If the overall statistics will show 0 failed tests at the end, then all is pretty fine.

**Comment by Vascom** [comment]
```
Launching unit test 'core.io.encdec'
--------------------------------------------------------------------------------
Unit test 'core.io.encdec' has failed, execution time: 5.28 s
--------------------------------------------------------------------------------
Overall unit test statistics:
  execution time [s]:   29.20
  launched:             88
  ignored:              0
  succeeded:            87
  failed:               1
```
--------------------------------------------------------------------------------
BUILDSTDERR: List of failed test+ RPM_EC=0

**Comment by sadko4u** [comment]
Now another test is failing. Do you have the test log for the last test?

**Comment by sadko4u** [comment]
Ah, probaby it's test overtime.

**Comment by sadko4u** [comment]
Please try this commit and save output to the log: 5ba34b76b2393045f68de4566107dbd887a380be.
I believe we should add more verbosity:
```
.test/lsp-plugins-test utest -v -j 1 >lsp-plugins-utest.log 2>&1
```

**Comment by Vascom** [comment]
```
Overall unit test statistics:
  execution time [s]:   7.58
  launched:             88
  ignored:              0
  succeeded:            88
  failed:               0
```

**Comment by Vascom** [comment]
Full test log.
[lsp-plugins-utest_3.log](https://github.com/sadko4u/lsp-plugins/files/3338969/lsp-plugins-utest_3.log)

**Comment by sadko4u** [comment]
OK then. I've backported all changes from **devel** branch to the **ppc64-hotfix** branch. Now try to build from this branch.

**Comment by Vascom** [comment]
Built and tests passed. But only 77 tests launched.
```
Overall unit test statistics:
  execution time [s]:   10.74
  launched:             77
  ignored:              0
  succeeded:            77
  failed:               0
```

**Comment by sadko4u** [comment]
> Built and tests passed. But only 77 tests launched.

That's normal since the **devel** branch is too far ahead of the **lsp-plugins-1.1.9** tag and **hotfix** branch.
Here's the final patch: 34dcdef597392019a605f3e4ecd7358f5ff19436
To build 1.1.9 version, you need checkout all patches from **hotfix** branch and apply them to the **lsp-plugins-1.1.9** tagged commit.

**Comment by Vascom** [comment]
I can wait 1.1.10.

**Comment by sadko4u** [comment]
> I can wait 1.1.10.

I believe it will last one or two months. Current development is still in progress and there are still a lot of things to do.

**Comment by Vascom** [comment]
OK.
Really I was needed only lv2 plugins for pulseeffects. So vst plugins for ppc64le can wait for the next release.

**Comment by sadko4u** [comment]
The problem is not only VST-related. So to build the 1.1.9 release the hotfix was really needed.

**Comment by sadko4u** [comment]
What's with s390x now?

**Comment by Vascom** [comment]
As I said - ppc64le is optional arch and can wait.
I don't want package many patches just for ppc64le support if will be new release soon.

**Comment by Vascom** [comment]
I can test s390x same way.

**Comment by sadko4u** [comment]
> I can test s390x same way.

Yes, it would be good.

**Comment by Vascom** [comment]
You will create new branch?

**Comment by sadko4u** [comment]
Pushed branch **s390x-port**, removed branch **ppc64-port**

**Comment by Vascom** [comment]
s390x tests log
[lsp-plugins-utest_s390x_0.log](https://github.com/sadko4u/lsp-plugins/files/3339343/lsp-plugins-utest_s390x_0.log)
```
  failed:               3
--------------------------------------------------------------------------------
List of failed tests:
  core.protocol.osc
  core.string
  core.io.charset
```

**Comment by sadko4u** [comment]
Hmmm. Need to watch:
```gcc -dM -E - < /dev/null```

**Comment by Vascom** [comment]
May be tomorrow or in monday.

**Comment by Vascom** [comment]
[s390x-gcc.log](https://github.com/sadko4u/lsp-plugins/files/3344120/s390x-gcc.log)

**Comment by sadko4u** [comment]
s390x is the first big-endian architecture for LSP Plugins. I've found that big-endianess test in header files produced invalid macro definitions that defined both *ARCH_LE* and *ARCH_BE* macros. Because *ARCH_LE* macros is used in most places as a base check of endianess, part of code considered the architecture as little-endian.
I've committed several changes to the branch, please test the build again.

**Comment by Vascom** [comment]
Build failed https://kojipkgs.fedoraproject.org//work/tasks/5721/35975721/build.log

**Comment by sadko4u** [comment]
Good. Here's the patch: 2bdb6818eec46e55e03ca83398c883d3e4602030

**Comment by Vascom** [comment]
Build OK but one test failed.
And wrong Architecture in test - ppc64.
[lsp-plugins-utest_s390x_1.log](https://github.com/sadko4u/lsp-plugins/files/3344472/lsp-plugins-utest_s390x_1.log)

**Comment by sadko4u** [comment]
Here are some fixes: 1c30feb1a72c4432558930339581adf9565b84cd

**Comment by Vascom** [comment]
Now all is OK.
```
Overall unit test statistics:
  execution time [s]:   8.11
  launched:             88
  ignored:              0
  succeeded:            88
  failed:               0
```

**Comment by sadko4u** [comment]
Merged changes into **devel** branch: 455f58c4ff7b9b5aa115dda2780cd71608a6dbf3
Backported changes to **s390x-port-hotfix**: 5d4dd7f6f52d407013d486b2c7503cea2b549638

Please try to build from **s390x-port-hotfix** now.

**Comment by Vascom** [comment]
Good.
```
Overall unit test statistics:
  execution time [s]:   7.55
  launched:             77
  ignored:              0
  succeeded:            77
  failed:               0
```

**Comment by sadko4u** [comment]
Merged changes into **hotfix** branch: 2ec26658d62b0c4b58f3b3572509bdb526604da1
Removed **s390x-port-hotfix** and **s390x-port** branches.

Seems that all we needed is now complete.
Well done!

**Comment by Vascom** [comment]
Build is good.

**Comment by sadko4u** [comment]
The changes will be available in 1.1.10 release.
I'll close this topic when the release will be shipped.

**Comment by Vascom** [comment]
Thanks.

**Comment by sadko4u** [comment]
@Vascom 
I'm currently preparing for the release.
Because I've a bit changed build process, I want to ask you again to build from **devel** branch and ensure that all builds OK.

**Comment by Vascom** [comment]
I will check it tomorrow.

**Comment by sadko4u** [comment]
@Vascom , how are the things going on?

**Comment by Vascom** [comment]
Built successful for all arches.

**Comment by sadko4u** [comment]
> Built successful for all arches.

Thank you for checks.

**Comment by sadko4u** [comment]
Available in 1.1.10 release!
